### PR TITLE
components/autocomplete: wip transition from class to functional react (for #1426)

### PR DESCRIPTION
**Pull request by matthewmueller**
**User-Facing Changes**

Nothing should change.

**Description**

<s>Not quite working yet, but I wanted to commit early to make sure it's going in the right direction. </s>
Refactored the Autocomplete component from a class-based React component to a hook-based functional React component.

**Todos**

- [x] Code ported over to a functional component
- [x] Get Storybook to work the same across the class and functional variants
- [x] Update `<Autocomplete>` usage across the codebase
- [x] Cleanup variables now that it's working (_onBlur => onBlur)
- [x] Make the linter happy
-  Anything else?

**Notes**

I've left some notes in comments below

**References**

This is part of https://github.com/foxglove/studio/issues/1426

### CODE REVIEW

File: packages/studio-base/src/panels/diagnostics/DiagnosticStatusPanel.tsx [hunk 1]
```diff
@@ -55,11 +55,7 @@ function DiagnosticStatusPanel(props: Props) {
   } = config;
 
   const onSelect = useCallback(
-    (
-      _value: string,
-      entry: DiagnosticAutocompleteEntry,
-      autocomplete: Autocomplete<DiagnosticAutocompleteEntry>,
-    ) => {
+    (_value: string, entry: DiagnosticAutocompleteEntry, autocomplete: AutocompleteAPI) => {
```
[the remark this answers is not in the document]
Comment (matthewmueller) [reply packages/studio-base/src/panels/diagnostics/DiagnosticStatusPanel.tsx]: I started adjusting this, but it ended up being more code than keeping it as is. I'd suggest revisiting if onSelect needs more params.

File: packages/studio-base/src/panels/ThreeDimensionalViz/FollowTFControl.tsx [hunk 2]
```diff
@@ -184,18 +186,26 @@ const FollowTFControl = memo<Props>((props: Props) => {
...
+    ],
   );
 
   const openFrameList = useCallback(() => {
     setForceShowFrameList(true);
-    if (autocomplete.current) {
-      autocomplete.current.focus();
+    if (autocomplete) {
+      autocomplete.focus();
     }
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDimensionalViz/FollowTFControl.tsx:209]: nit for future (not important enough to block this merging): can use TS optional chaining `autocomplete?.focus()`

File: packages/studio-base/src/panels/ThreeDimensionalViz/FollowTFControl.tsx [hunk 3]
```diff
@@ -147,7 +149,7 @@ const FollowTFControl = memo<Props>((props: Props) => {
   const nodesWithoutDefaultFollowTfFrame = allNodes?.length;
   const newFollowTfFrame = allNodes?.[0]?.tf?.id;
 
-  const autocomplete = createRef<Autocomplete<TfTreeNode>>();
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDimensionalViz/FollowTFControl.tsx:150]: Was this change required? I think it's fine, just wondering why the change was made since I would expect it not to be necessary.

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 4]
```diff
@@ -181,308 +182,298 @@ const HighlightChars = (props: { str: string; indices: Set<number> }) => {
...
+  const [stateValue, setValue] = useState<string | undefined>(undefined);
+  const [focused, setFocused] = useState<boolean>(false);
+  const [stateIgnoreBlur, setIgnoreBlur] = useState<boolean>(false);
+
+  // Props
+  const {
+    autocompleteKey,
+    autoSize = false,
+    getItemText = defaultGetText("getItemText"),
+    getItemValue = defaultGetText("getItemValue"),
```
Comment (jtbandes) [inline packages/studio-base/src/components/Autocomplete.tsx:200]: These are now being created on each render where previously they were created once when building the `static defaultProps`. Might be nice to do these with useMemo instead, e.g. `const getItemText = useMemo(() => props.getItemText ?? defaultGetText("getItemText"), [props.getItemText])`

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 5]
```diff
@@ -181,308 +182,298 @@ const HighlightChars = (props: { str: string; indices: Set<number> }) => {
...
+    sortWhenFiltering = true,
+    clearOnFocus = false,
+    minWidth = 100,
+    menuStyle = {},
+    inputStyle = {},
+    onChange: onChangeCallback,
+    onSelect: onSelectCallback,
+    onBlur: onBlurCallback,
+    disableAutoSelect,
+    ref = () => {},
```
Comment (jtbandes) [inline packages/studio-base/src/components/Autocomplete.tsx:215]: Does this actually work? ref can't normally passed via props, and can only be accessed in functional components by wrapping with `forwardRef`. At least that's my understanding.

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 6]
```diff
@@ -181,308 +182,298 @@ const HighlightChars = (props: { str: string; indices: Set<number> }) => {
...
+  }: AutocompleteProps<T> = props;
+
+  const autocompleteItems: FzfResultItem<T>[] = filterText
+    ? new Fzf(items, {
+        fuzzy: filterText.length > 2 ? "v2" : false,
+        sort: sortWhenFiltering,
+        limit: MAX_ITEMS,
+        selector: getItemText as (_: unknown) => string, // Fzf selector TS type seems to be wrong?
+      }).find(filterText)
+    : items.map((item) => itemToFzfResult(item));
```
Comment (jtbandes) [inline packages/studio-base/src/components/Autocomplete.tsx:225]: We'll want this wrapped in a useMemo so we don't call .find() unnecessarily if it's expensive. I'm not sure if the initialization of `new Fzf` is expensive either, but could consider memoizing that separately.

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 7]
```diff
@@ -181,308 +182,298 @@ const HighlightChars = (props: { str: string; indices: Set<number> }) => {
...
-  }
+  }, [onBlurCallback]);
 
-  private _onFocus = (): void => {
-    if (this._ignoreFocus) {
-      return;
-    }
-    const { clearOnFocus } = this.props;
+  // Give callers an opportunity to control autocomplete
+  ref({ setSelectionRange, focus, blur });
```
Comment (jtbandes) [inline packages/studio-base/src/components/Autocomplete.tsx:263]: useImperativeHandle

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 8]
```diff
@@ -181,308 +182,298 @@ const HighlightChars = (props: { str: string; indices: Set<number> }) => {
...
+        onChangeCallback(event, (event.target as HTMLInputElement).value);
+      } else {
+        setValue((event.target as HTMLInputElement).value);
+      }
+    },
+    [onChangeCallback],
+  );
+
+  // Make sure the input field gets focused again after selecting, in case we're doing multiple
+  // autocompletes. We pass in `this` to `onSelect` in case the user of this component wants to call
```
Comment (jtbandes) [inline packages/studio-base/src/components/Autocomplete.tsx:277]: nit: `this` comment is out of date

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 9]
```diff
@@ -109,21 +116,21 @@ const classes = mergeStyleSets({
 // it doesn't cause any problems.)
 type AutocompleteProps<T = unknown> = {
   items: T[];
-  getItemValue: (arg0: T) => string;
-  getItemText: (arg0: T) => string;
+  getItemValue?: (arg0: T) => string;
+  getItemText?: (arg0: T) => string;
```
[the remark this answers is not in the document]
Comment (jtbandes) [reply packages/studio-base/src/components/Autocomplete.tsx:114]: Anything present in `static defaultProps` is allowed to be non-optional in Props.

File: packages/studio-base/src/panels/ThreeDimensionalViz/FollowTFControl.tsx (hunk 3 above)
Comment (matthewmueller) [reply packages/studio-base/src/panels/ThreeDimensionalViz/FollowTFControl.tsx:150]: I changed this because I wasn't able to figure out a way to manually write to a ref done here: https://github.com/foxglove/studio/pull/1986/files/49a7c781291d20572f93c7e76c62a6555c54b001#diff-0b248d633e53e623ebd68e8998751ce04b38d2801d3e2d64b2cd5602b25c39deR257

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 5 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx:215]: This does work, but I'm not sure it's the right way to do things.

My understanding is that ref is used exclusively to reference DOM nodes. The ref in this case is just a JS object that contains some methods for controlling autocomplete, allow other components to trigger re-renders. This is why I originally called this useAPI or something other than ref to avoid confusion.

Is there a way to use refs for non-DOM nodes? If so, is there some documentation I could take a look at or do we have any examples in the codebase?

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 4 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx:200]: Good catch! Fixed.

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 7 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx:263]: See https://github.com/foxglove/studio/pull/1986#discussion_r742531416. I'd like to clear that up before making any changes here.

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 6 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx:225]: Good call! Still learning all the various hooks. I appreciate for the reminders. I'll most likely need a few more of these 🙂

Regarding also memoizing `Fzf`. I don't think it makes much sense in this case because `Fzf` depends on `filterText` all the same, so every time `filterText` changes, you'd want to re-initialize `Fzf`. It's a rather bad use of a class actually.

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 5 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx]: Actually I just started seeing errors in development about passing a ref through. Odd that I didn't see it before. I just adjusted it to `controller` for the time-being. If there's a better way to pass non-DOM nodes as refs, please let me know!

File: packages/studio-base/src/panels/ThreeDimensionalViz/FollowTFControl.tsx [hunk 10]
```diff
@@ -248,7 +254,7 @@ const FollowTFControl = memo<Props>((props: Props) => {
     >
       {showFrameList && (
         <Autocomplete
-          ref={autocomplete}
+          controller={(api) => (autocomplete = api)}
```
Comment (matthewmueller) [inline packages/studio-base/src/panels/ThreeDimensionalViz/FollowTFControl.tsx:257]: renamed because ref is special

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 5 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx]: **Update** I just renamed it from ref to controller.

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 11]
```diff
@@ -181,336 +187,319 @@ const HighlightChars = (props: { str: string; indices: Set<number> }) => {
...
+            sort: sortWhenFiltering,
+            limit: MAX_ITEMS,
+            selector: getItemText as (_: unknown) => string, // Fzf selector TS type seems to be wrong?
+          }).find(filterText)
+        : items.map((item) => itemToFzfResult(item)),
+    [filterText, getItemText, items, sortWhenFiltering],
+  );
+
+  const hasError =
+    props.hasError === true ? props.hasError : autocompleteItems.length === 0 && value?.length;
```
Comment (jtbandes) [inline packages/studio-base/src/components/Autocomplete.tsx:250]: I think this logic changed unintentionally. Looking at the storybook screenshots, I see red/error text in a lot more cases than before. The original behavior was that the computed value based on autocompleteItems.length would be used only if props.hasError was undefined.

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 5 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx]: Ah, TIL. Thanks!

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 11 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx]: Great catch, 

- in main: if `props.hasError` is false, then `hasError` would be false
- in this branch: if `props.hasError` is false, then `hasError` is computed from the autocomplete items.

Where are the storybook screenshots? 

I took a look at [this branch](https://www.chromatic.com/component?buildNumber=4416&historyLengthAtIndex=33&distanceToMoveBack=-22&appId=603ec8bf7908b500231841e2&name=components%2FAutocomplete&specName=with+a+long+truncated+path+%28and+autoSize%29&componentInspectorKey=6183fd02de17a7003afe1f56-1200-interactive-true) vs. [main](https://www.chromatic.com/component?appId=603ec8bf7908b500231841e2&name=components%2FAutocomplete&buildNumber=4446&componentInspectorKey=6185b774a34691003aa092dd-1200-interactive-true&historyLengthAtIndex=5&distanceToMoveBack=-1) but didn't notice a difference.

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 12]
```diff
@@ -246,9 +246,7 @@ export default React.forwardRef(function Autocomplete<T = unknown>(
     [filterText, getItemText, items, sortWhenFiltering],
   );
 
-  const hasError =
-    props.hasError === true ? props.hasError : autocompleteItems.length === 0 && value?.length;
-
+  const hasError = props.hasError ?? (autocompleteItems.length === 0 && value?.length);
```
Comment (matthewmueller) [inline packages/studio-base/src/components/Autocomplete.tsx:249]: If props.hasError is null or undefined, then compute from autocompleteItems. Since hasError is typed only to undefined | boolean, it should be the same behavior as main.

I was tempted to just use the old school, `typeof props.hasError === 'undefined'`, but took this opportunity to refresh on the new syntax (I like it!)

**Comment by matthewmueller** [comment]
FYI, incorporated https://github.com/foxglove/studio/pull/2078#pullrequestreview-798200186 into this branch.

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 11 above)
Comment (jtbandes) [reply packages/studio-base/src/components/Autocomplete.tsx]: Not sure about the Autocomplete stories in particular but Autocomplete ends up getting used in lots of different stories. Here's an example of a build from this branch showing 88 changed stories; I think I was looking at something like this (but not exactly sure which build it was): https://www.chromatic.com/build?appId=603ec8bf7908b500231841e2&number=4416 But now that you've made fixes, the number of changes has gone down to 1: 
image

I think Chromatic keeps track of intermediate approvals so unfortunately it's not always possible to get a complete inventory of what's changed from main to the current branch head...

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 13]
```diff
@@ -181,336 +187,317 @@ const HighlightChars = (props: { str: string; indices: Set<number> }) => {
...
+            limit: MAX_ITEMS,
+            selector: getItemText as (_: unknown) => string, // Fzf selector TS type seems to be wrong?
+          }).find(filterText)
+        : items.map((item) => itemToFzfResult(item)),
+    [filterText, getItemText, items, sortWhenFiltering],
+  );
+
+  const hasError = props.hasError ?? (autocompleteItems.length === 0 && value?.length);
+  const open = focused && autocompleteItems.length > 0;
+  const ignoreBlur = open ? stateIgnoreBlur : false;
```
Comment (jtbandes) [inline packages/studio-base/src/components/Autocomplete.tsx:251]: Not sure I totally get what this is for — is this a change from the original implementation?

**Review by jtbandes — APPROVED** [review]

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 11 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx]: Ah, I see that now. That's super nice. Okay so next time I'll check out the Unreviewed changes in Chromatic. Thanks!

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 13 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx:251]: Good catch! Yes, but it wasn't intentional :-) 

This translation would trigger re-renders, which I could see causing side-effects when you're considering event order (blur coming before click).

I adjusted the implemented to use references, which shouldn't trigger a re-render. I wasn't able to confirm the problem with the existing code though. It seems like it's most obvious with multiple autocompletes. I took a look through the codebase, but couldn't find an example of this. The comment mentions the Plot which doesn't seem to depend on Autocomplete as far as I can tell.

File: packages/studio-base/src/components/Autocomplete.tsx [hunk 14]
```diff
@@ -181,336 +187,322 @@ const HighlightChars = (props: { str: string; indices: Set<number> }) => {
...
+  // References
+  const autocomplete = useRef<ReactAutocomplete>(ReactNull);
+  const ignoreBlur = useRef<boolean>(false);
 
-class AutocompleteImpl<T = unknown>
-  extends PureComponent<AutocompleteProps<T>, AutocompleteState>
-  implements IAutocomplete
-{
-  private _autocomplete: RefObject<ReactAutocomplete>;
-  private _ignoreFocus: boolean = false;
```
Comment (matthewmueller) [inline packages/studio-base/src/components/Autocomplete.tsx:195]: Note that I removed `_ignoreFocus` because it seemed like unused logic. It's never set to true, so the conditional in `onFocus` always passes.

File: packages/studio-base/src/components/Autocomplete.tsx (hunk 13 above)
Comment (matthewmueller) [reply packages/studio-base/src/components/Autocomplete.tsx]: Maybe we add a story that demonstrates this problem?

Comment (jtbandes) [reply packages/studio-base/src/components/Autocomplete.tsx]: If we can't easily figure out how to reproduce but we are keeping the existing behavior, I think it's probably not worth the effort to add a story for it right now. I think we are eventually due for a larger rewrite of this component because react-autocomplete is no longer maintained...
